packages/Discretization/src/DTK_PointSearch_def.hpp

```diff
@@ -0,0 +1,736 @@
+            DTK_MARK_REGION( "fill_block_cell_" + std::to_string( topo_id ) ),
+            Kokkos::RangePolicy<ExecutionSpace>( 0, n_cells ),
+            KOKKOS_LAMBDA( int const i ) {
+                if ( cell_topologies( i ) == topo_id )
+                {
+                    internal::computeBlockCellsBoundingBox(
+                        dim, i, n_nodes_per_topo( topo_id ), node_offset( i ),
+                        topo_id, cells, offset, coordinates, block_cells,
+                        bounding_boxes, bounding_box_to_cell );
+                }
+            } );
+        Kokkos::fence();
+    }
+}
+
+template <typename DeviceType>
+void PointSearch<DeviceType>::performDistributedSearch(
+    Kokkos::View<double **, DeviceType> points_coord,
+    Kokkos::View<Box *, DeviceType> bounding_boxes,
+    Kokkos::View<Point *, DeviceType> &imported_points,
+    Kokkos::View<int *, DeviceType> &imported_query_ids,
+    Kokkos::View<int *, DeviceType> &imported_cell_indices,
+    Kokkos::View<int *, DeviceType> &ranks )
+{
+    DistributedSearchTree<DeviceType> distributed_tree( _comm, bounding_boxes );
+
+    unsigned int const n_points = points_coord.extent( 0 );
+
+    // Build the queries
+    // FIXME do not use Within predicate because it requires a radius, i.e., it
+    // is mesh dependent
+    using ExecutionSpace = typename DeviceType::execution_space;
+    Kokkos::View<Details::Within *, DeviceType> queries( "queries", n_points );
+    Kokkos::parallel_for(
+        "register_queries", Kokkos::RangePolicy<ExecutionSpace>( 0, n_points ),
+        KOKKOS_LAMBDA( int i ) {
+            queries( i ) =
+                Details::within( {{points_coord( i, 0 ), points_coord( i, 1 ),
+                                   points_coord( i, 2 )}},
+                                 1e-14 );
+        } );
+    Kokkos::fence();
+
+    // Perform the distributed search
+    Kokkos::View<int *, DeviceType> indices( "indices" );
+    Kokkos::View<int *, DeviceType> offset( "offset" );
+    distributed_tree.query( queries, indices, offset, ranks );
+
+    // Create the source to target distributor
+    auto ranks_host = Kokkos::create_mirror_view( ranks );
+    Kokkos::deep_copy( ranks_host, ranks );
+    Tpetra::Distributor source_to_target_distributor( _comm );
+    unsigned int const n_imports = source_to_target_distributor.createFromSends(
+        Teuchos::ArrayView<int const>( ranks_host.data(),
+                                       ranks_host.extent( 0 ) ) );
+
+    // Communicate cell indices
+    Kokkos::realloc( imported_cell_indices, n_imports );
+    source_to_target_distributor.doPostsAndWaits( indices, 1,
+                                                  imported_cell_indices );
+    // Duplicate the points_coord for the communication. Duplicating the points
+    // allows us to use the same distributor.
+    unsigned int const indices_size = indices.extent( 0 );
+    Kokkos::View<Point *, DeviceType> exported_points( "exported_points",
```

Comment:
If `Point` has `Teuchos::SerializationTraits` do we need to do this duplication?

Reply:
It has nothing to do with serialization. This is because of the output format of the search.